src/Workspaces/Core/Portable/Execution/AssetBuilder.cs

```diff
@@ -24,112 +24,117 @@ public AssetBuilder(IChecksumTreeNode checksumTree)
+        public Asset Build(ProjectState projectState, ParseOptions parseOptions, CancellationToken cancellationToken)
         {
-            return _checksumTree.GetOrCreateAssetAsync(parseOptions, projectState, WellKnownChecksumObjects.ParseOptions, CreateParseOptionsAsync, cancellationToken);
+            return _checksumTree.GetOrCreateAsset(parseOptions, projectState, WellKnownChecksumObjects.ParseOptions, CreateParseOptions, cancellationToken);
         }
 
-        public Task<Asset> BuildAsync(ProjectReference reference, CancellationToken cancellationToken)
+        public Asset Build(ProjectReference reference, CancellationToken cancellationToken)
         {
-            return _checksumTree.GetOrCreateAssetAsync(reference, reference, WellKnownChecksumObjects.ProjectReference, CreateProjectReferenceAsync, cancellationToken);
+            return _checksumTree.GetOrCreateAsset(reference, reference, WellKnownChecksumObjects.ProjectReference, CreateProjectReference, cancellationToken);
         }
 
-        public Task<Asset> BuildAsync(MetadataReference reference, CancellationToken cancellationToken)
+        public Asset Build(MetadataReference reference, CancellationToken cancellationToken)
         {
-            return _checksumTree.GetOrCreateAssetAsync(reference, reference, WellKnownChecksumObjects.MetadataReference, CreateMetadataReferenceAsync, cancellationToken);
+            return _checksumTree.GetOrCreateAsset(reference, reference, WellKnownChecksumObjects.MetadataReference, CreateMetadataReference, cancellationToken);
         }
 
-        public Task<Asset> BuildAsync(AnalyzerReference reference, CancellationToken cancellationToken)
+        public Asset Build(AnalyzerReference reference, CancellationToken cancellationToken)
         {
-            return _checksumTree.GetOrCreateAssetAsync(reference, reference, WellKnownChecksumObjects.AnalyzerReference, CreateAnalyzerReferenceAsync, cancellationToken);
+            return _checksumTree.GetOrCreateAsset(reference, reference, WellKnownChecksumObjects.AnalyzerReference, CreateAnalyzerReference, cancellationToken);
         }
 
-        public Task<Asset> BuildAsync(TextDocumentState state, SourceText unused, CancellationToken cancellationToken)
+        public Asset Build(TextDocumentState state, SourceText text, CancellationToken cancellationToken)
         {
             // TODO: currently this is a bit wierd not to hold onto source text.
             //       it would be nice if SourceText is changed like how recoverable syntax tree work.
-            return _checksumTree.GetOrCreateAssetAsync(state, state, WellKnownChecksumObjects.SourceText, CreateSourceTextAsync, cancellationToken);
+            var asset = _checksumTree.GetOrCreateAsset(state, state, WellKnownChecksumObjects.SourceText, CreateSourceText, cancellationToken);
+
+            // make sure we keep text alive. this is to make sure we don't do any async call in asset builder
+            GC.KeepAlive(text);
```

Comment:
I don't want to create lambda or create new interface just for this. I need a way to get rid of text being async anyway though.